### ISSUE TITLE
Highlighting of member implementations/overrides

### ISSUE THREAD

**Issue by blaugold** [Dart-Code/Dart-Code#4670]
When I place the cursor at a type, other references to it in the same file are highlighted (including in the scroll bar):

<img width="518" alt="Screenshot 2023-07-31 at 22 28 24" src="https://github.com/Dart-Code/Dart-Code/assets/5929595/415fdf62-b6a4-4ea8-946f-eb66449bc184">

I would like to have the option to enable the same behavior for members of a type:

<img width="518" alt="Screenshot 2023-07-31 at 22 28 20" src="https://github.com/Dart-Code/Dart-Code/assets/5929595/8ba9f0a3-22fb-4b95-bd72-74b28bdaca69">

**Comment by DanTup** [comment]
@bwilkerson I was surprised this didn't already happen (since a rename or Find Reference would find the overrides). I'm not sure if it was deliberate or just a consequence of grouping by Elements and not looking at the hierarchy. I tested in TypeScript and it does include members from subclasses.

I thought perhaps the occurrences could be grouped by their base/super elements, although I'm not sure if that would always result in the same items in each group depending on where you start. For example in:

```dart

abstract class A {
  void a();
}

mixin M {
  void a() {}
}

class B extends A with M {
  @override
  void a() {}
}
```

If we put the cursor here in `B.a()` you might reasonably expect that `A.a()` and `M.a()` would be highlighted too. However if you start from `M.a()` it seems less clear that `A.a()` should be highlighted (especially if `B` was in another file). Having a different set of things highlighted depending on which occurrence the text cursor was in doesn't seem right though.

Any thoughts/opinions?

**Comment by blaugold** [comment]
My intuition is that the highlighting would represent sub/superclass relationships (also seems that's how it works in VS Code for TypeScript). It would feel strange to me if putting the cursor on `M.a` would also highlight `A.a`.

**Comment by DanTup** [comment]
But what if you put the cursor in `B.a()`? I feel like I'd expect both `M.a()` and `A.a()` to be highlighted (after all, they'd both be renamed if you renamed `B.a()`). However the way this currently works (at least for Dart) is that these occurrences are groups - so each member of the group highlights every other member of the group - everything is symmetrical.

Something I just noticed is given the  code above, if you rename `B.a()`, it renames all three `a()`s. However if you rename `M.a()` it only renames `M.a()` and `B.a()` (leaving an error that B doesn't implement abstract `A.a()`). So renames are definitely not symmetrical. I'm wondering whether document highlights should be the same (and highlight the things that would be renamed if you renamed that symbol). This would be a change from the behaviour today though.

**Comment by blaugold** [comment]
> But what if you put the cursor in B.a()? I feel like I'd expect both M.a() and A.a() to be highlighted

Yes, I think so too! 

I meant to say that I would find it strange if members are highlighted that don't have an inheritance relationship to the member at the cursor location. When the cursor is at `B.a`, there is an inheritance relationship to `M.a` and `A.a`, but not to `A.a` when the cursor is at `M.a`.

**Comment by bwilkerson** [comment]
> I'm not sure if it was deliberate or just a consequence ...

It was just a consequence of the implementation, which, as you noted, doesn't look at the hierarchy. The reason the implementation doesn't look at the hierarchy is unknown (to me), but one possibility is the cost of looking for overridden members whenever a member is selected. I don't know what the performance impact would be of such a change, but I'd want to measure it before committing to such a change.

Other than a possible performance concern, I don't have any problem changing the behavior, especially if that's how `textDocument/documentHighlight` is expected to work.

**Comment by DanTup** [comment]
I'm wondering if we should create a separate implementation for LSP. The current implementation walks the whole file and groups all occurrences by their element (or a canonical element, which resolves getters). For LSP we only use one of these groups to satisfy the request.

I'm wondering if one way to improve this (without searching out for overrides) would be to canonicalize each element to the super-most declaration (perhaps via `InheritanceManager3`?). Doing that for the entire file is more expensive than we need it to be, because we only care about other things in a specific group, we could skip the work for symbols that clearly won't match (for example if the cursor is on `foo`, I believe it'll never match any other symbols not named `foo`).

So, could we:

- Get the name of the element at the cursor
- Resolve it to the super-most declaration
- Walk the document, and for only symbols of the same name*, resolve them to the super-most declaration and compare it to the previous one (if it's a match, include it in the group)

\* maybe we can do more than first check the name, but the goal is to avoid looking up inheritance information for symbols that definitely won't match

I don't know how expensive this would be so we should definitely measure it, but I'm interested in your opinion of whether this sounds sensible. It might also give is a route to fix https://github.com/Dart-Code/Dart-Code/issues/4454 for LSP that supports `.new` and only exclude it from the legacy protocol (I made comments about this issues fixing this at https://github.com/Dart-Code/Dart-Code/issues/4454#issuecomment-1649933953).

**Comment by bwilkerson** [comment]
Yes, I think it makes sense to try that experiment.

**Comment by DanTup** [comment]
I was starting to look at this, and above I said:

> Having a different set of things highlighted depending on which occurrence the text cursor was in doesn't seem right though.

And we'd discussed that [VS Code doesn't correctly handle asymmetric highlights](https://github.com/microsoft/vscode/issues/238301) before.

However today I discovered that TypeScript is also producing asymmetric highlights in this case (and therefore is also affected by VS Code's optimisation):

https://github.com/user-attachments/assets/e87c9625-1e01-47e5-85db-c320cda11857

But this makes me think that maybe the asymmetric highlights is the right call (and perhaps in some other places, like `getter/setter` with `foo++` that I just raised. WDYT @bwilkerson? I feel like the VS Code issue would be a trivial change, so if we think we should go down this route, I could try a PR for that.

**Comment by bwilkerson** [comment]
Based on the research I've done, asymmetry can be fairly harmful, leading users to be confused by and even mistrust the tools, so I'm reluctant to use it without significant justification. It might be the best solution in cases where the language effectively forces it (such as identifiers with two meanings), but isn't necessarily the right choice everywhere.

> I'm wondering if one way to improve this (without searching out for overrides) would be to canonicalize each element to the super-most declaration (perhaps via InheritanceManager3?).

I missed this before, but I don't think this will work. We resolve members to the first member found starting with the static type of the receiver. That might not match the super-most declaration.

What might work is, given a member that was selected, get the element for the enclosing type. Then, for every class-like declaration, look to see whether it defines a member of the same name, and if so whether one of the two types is a subtype of the other. That would allow us to highlight both overridden members as well as overriding members, and I think it would be (at least mostly) symmetric.

**Comment by DanTup** [comment]
> It might be the best solution in cases where the language effectively forces it (such as identifiers with two meanings), but isn't necessarily the right choice everywhere.

I'm not sure if this means you're in favour of it or not here 🙂


> What might work is, given a member that was selected, get the element for the enclosing type. Then, for every class-like declaration, look to see whether it defines a member of the same name, and if so whether one of the two types is a subtype of the other. That would allow us to highlight both overridden members as well as overriding members, and I think it would be (at least mostly) symmetric.

I'm not sure if I understand correctly - if I understand correctly I think this would be less symmetrical than I had in mind. Consider:

```dart
class A {
  aaa() {} // 1
}

class B extends A {
  aaa() {} // 2
}

class C extends A {
  aaa() {} // 3
}
```

If I put my text cursor in `A.aaa` (1), I think with your rules then 1+2+3 would be highlighted. But if I started in `C.aaa`, only 1+3 would?

Whereas my idea had been to try and get back to `A.aaa` if you started at `B.aaa` or `C.aaa` so that they were all highlighted symmetrically.


Another example is this:

```dart
abstract class A {
  aaa() {} // 1
}

interface class I {
  aaa() {} // 2
}

mixin M1 {
  aaa() {} // 3
}

mixin M2 on A {
  aaa() {} // 4
}

class B extends A with M1, M2 implements I {
  aaa() {} // 5
}
```

What's highlighted if I put my text cursor at `B.aaa` (5)? And is this different to if I started at `A.aaa` (1)?

**Comment by bwilkerson** [comment]
Yes, I realized last night that what I had written wouldn't be symmetric.

> I'm not sure if this means you're in favour of it or not here

It means I'm undecided and trying to think through the choices. I'll try to be more clear when I'm just thinking out loud.

> Whereas my idea had been to try and get back to A.aaa if you started at B.aaa or C.aaa so that they were all highlighted symmetrically.

That's what I thought you were suggesting. And that works for the overrides, but not for the references. Consider
```dart
class A {
  aaa() {} // 1
}

class B extends A {
  aaa() {} // 2
}

class C extends A {
  aaa() {} // 3
}

void f(A a, B b, C c) {
  a.aaa();
  b.aaa();
  c.aaa();
}
```

If the user places the cursor at 1, and if we then highlight all three declarations, then we probably want all three of the invocations to be highlighted, but each invocation resolves to a different method, so if we only use `A.aaa` as the element we're looking for we'll miss some. What we really need to do is find all of the overridden and overriding methods we want to highlight and then change the highlighter to look for references to multiple elements.

**Comment by DanTup** [comment]
> so if we only use `A.aaa` as the element we're looking for we'll miss some.

I was thinking something like this:

- You put your cursor at (2) (`B.aaa`)
- We walk up the hierarchy to find that the super-most declaration of `aaa` is `A.aaa` and consider that our target
- We walk down the AST, and for those invocations (if the name matches?), we also look up the super-most declaration for them, and compare that to the target

I haven't tried to implement this so I don't know how feasible it is. But in my mind, it _feels like_ the behaviour I expect, which is to look at the entire tree/hierarchy for all the branches (members) that match the branch (member) I'm on.


> What we really need to do is find all of the overridden and overriding methods we want to highlight and then change the highlighter to look for references to multiple elements.

Assuming we go up to the root and _then_ find all the overriding methods, I think this would work the same (we'd just start with the full set of targets instead of computing them)? But it we only take one pass "up the tree" then I think we'd miss "sibling" members?

**Comment by bwilkerson** [comment]
Yes, that would probably work better. I think it would handle cases where some of the classes in the hierarchy are defined outside the current file, which my idea probably wouldn't.

**Comment by DanTup** [comment]
Fixed by https://github.com/dart-lang/sdk/commit/2bf4d20cd3daa70b19aeb67349a6e078c7eb412d

The change is in the Dart analysis server which ships in the Dart/Flutter SDKs, so it'll show up in a future SDK release (not a Dart-Code release).
